File: csharp/src/Subscriber.cs
using MQTTnet;
using MQTTnet.Client;
using System.Diagnostics;
using MessagePack;
using PeterO.Cbor;
using Google.Protobuf;

namespace MQTTComparison
{
    public class Subscriber
    {
        private readonly IMqttClient _client;
        private readonly string _broker;
        private readonly int _port;
        private readonly string _topic;
        private readonly int _qos;
        private readonly string _encoding;
        private int _messageCount;
        private readonly List<double> _receiveTimes;

        public Subscriber(string broker, int port, string topic, int qos, string encoding = "json")
        {
            _broker = broker;
            _port = port;
            _topic = topic;
            _qos = qos;
            _encoding = encoding;
            _messageCount = 0;
            _receiveTimes = new List<double>();
            
            var factory = new MqttFactory();
            _client = factory.CreateMqttClient();
        }

        public async Task ConnectAsync()
        {
            var options = new MqttClientOptionsBuilder()
                .WithTcpServer(_broker, _port)
                .WithClientId("csharp_subscriber")
                .WithKeepAlivePeriod(TimeSpan.FromSeconds(60))
                .WithCleanSession(true)
                .Build();

            await _client.ConnectAsync(options);
            Console.WriteLine($"✓ Connected to {_broker}:{_port}");
        }

        public async Task SubscribeAsync()
        {
            _client.UseApplicationMessageReceivedHandler(OnMessageReceived);
            
            await _client.SubscribeAsync(new MqttTopicFilterBuilder()
                .WithTopic(_topic)
                .WithQualityOfServiceLevel((MQTTnet.Protocol.MqttQualityOfServiceLevel)_qos)
                .Build());
                
            Console.WriteLine($"✓ Subscribed to topic: {_topic} (QoS: {_qos})");
            Console.WriteLine("\nWaiting for messages (Ctrl+C to exit)...\n");
        }

        public async Task DisconnectAsync()
        {
            await _client.DisconnectAsync();
            Console.WriteLine("✓ Disconnected");
        }

        private SensorData DecodeMessage(byte[] payload)
        {
            return _encoding.ToLower() switch
            {
                "json" => JsonSerializer.Deserialize<SensorData>(System.Text.Encoding.UTF8.GetString(payload))!,
                "msgpack" => MessagePackSerializer.Deserialize<SensorData>(payload),
                "cbor" => CBORObject.DecodeFromBytes(payload).ToObject<SensorData>(),
                "protobuf" => JsonSerializer.Deserialize<SensorData>(System.Text.Encoding.UTF8.GetString(payload))!, // Fallback to JSON for now
                _ => throw new ArgumentException($"Unsupported encoding: {_encoding}")
            };
        }

        private async Task OnMessageReceived(MqttApplicationMessageReceivedEventArgs e)
        {
            var receiveTime = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            _messageCount++;

            try
            {
                var data = DecodeMessage(e.ApplicationMessage.Payload);

                Console.WriteLine($"\n[Message {_messageCount}] Topic: {e.ApplicationMessage.Topic}");
                Console.WriteLine($"  Sensor ID: {data.SensorId}");
                Console.WriteLine($"  Temperature: {data.Temperature:F2}°C");
                Console.WriteLine($"  Humidity: {data.Humidity:F2}%");
                Console.WriteLine($"  Pressure: {data.Pressure:F2} hPa");
                Console.WriteLine($"  Timestamp: {data.Timestamp:F6}");

                // Calculate receive latency
                var latency = receiveTime - data.Timestamp;
                _receiveTimes.Add(latency);
                Console.WriteLine($"  Receive latency: {latency * 1000:F2}ms");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"\n[Message {_messageCount}] Failed to decode message: {ex.Message}");
            }

            await Task.CompletedTask;
        }

        public void PrintSummary()
        {
            Console.WriteLine($"\n✓ Received {_messageCount} messages");
            
            if (_receiveTimes.Count > 0)
            {
                var avgLatency = _receiveTimes.Average();
                Console.WriteLine($"✓ Average receive latency: {avgLatency * 1000:F2}ms");
            }
        }
    }

    public class SubscriberProgram
    {
        private static Subscriber? _subscriber;

        public static async Task Main(string[] args)
        {
            // Parse command line arguments
            var options = ParseArgs(args);
            
            Console.WriteLine("=== MQTT Subscriber (C#) ===");
            Console.WriteLine($"Broker: {options.Broker}:{options.Port}");
            Console.WriteLine($"Topic: {options.Topic}");
            Console.WriteLine($"QoS: {options.Qos}");
            Console.WriteLine($"Encoding: {options.Encoding}");
            Console.WriteLine();

            // Set up signal handler
            Console.CancelKeyPress += (sender, e) =>
            {
                e.Cancel = true;
                _subscriber?.PrintSummary();
                _subscriber?.DisconnectAsync().Wait();
                Environment.Exit(0);
            };

            try
            {
                _subscriber = new Subscriber(options.Broker, options.Port, options.Topic, options.Qos, options.Encoding);
                await _subscriber.ConnectAsync();
                await _subscriber.SubscribeAsync();

                // Keep running until interrupted
                while (true)
                {
                    await Task.Delay(100);
                }
            }
            catch (Exception e)
            {
                Console.WriteLine($"✗ Error: {e.Message}");
                Environment.Exit(1);
            }
        }

        private static SubscriberOptions ParseArgs(string[] args)
        {
            var options = new SubscriberOptions();
            
            for (int i = 0; i < args.Length; i++)
            {
                switch (args[i])
                {
                    case "--broker" when i + 1 < args.Length:
                        options.Broker = args[++i];
                        break;
                    case "--port" when i + 1 < args.Length:
                        options.Port = int.Parse(args[++i]);
                        break;
                    case "--topic" when i + 1 < args.Length:
                        options.Topic = args[++i];
                        break;
                    case "--qos" when i + 1 < args.Length:
                        options.Qos = int.Parse(args[++i]);
                        break;
                    case "--encoding" when i + 1 < args.Length:
                        options.Encoding = args[++i];
                        break;
                }
            }
            
            return options;
        }
    }

    public class SubscriberOptions
    {
        public string Broker { get; set; } = "localhost";
        public int Port { get; set; } = 1883;
        public string Topic { get; set; } = "mqtt-demo/all";
        public int Qos { get; set; } = 1;
        public string Encoding { get; set; } = "json";
    }
}

File: rust/src/subscriber.rs
use clap::Parser;
use rumqttc::{Client, MqttOptions, QoS, Event, Incoming};
use serde::{Deserialize, Serialize};
use serde_json;
use rmp_serde;
use ciborium;
use prost;
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Parser)]
#[command(name = "mqtt-subscriber")]
#[command(about = "MQTT Subscriber for Rust with multiple encoding support")]
struct Args {
    #[arg(long, default_value = "localhost")]
    broker: String,
    
    #[arg(long, default_value = "1883")]
    port: u16,
    
    #[arg(long, default_value = "json")]
    encoding: String,
    
    #[arg(long, default_value = "mqtt-demo/all")]
    topic: String,
    
    #[arg(long, default_value = "1")]
    qos: u8,
}

#[derive(Serialize, Deserialize, Debug)]
struct SensorData {
    timestamp: f64,
    sensor_id: String,
    temperature: f64,
    humidity: f64,
    pressure: f64,
    location: Option<Location>,
    status: Option<String>,
    battery_level: Option<f64>,
    signal_strength: Option<i32>,
    sensor_readings: Option<Vec<f64>>,
    metadata: Option<Metadata>,
    additional_data: Option<String>,
}

#[derive(Serialize, Deserialize, Debug)]
struct Location {
    lat: f64,
    lon: f64,
    altitude: f64,
}

#[derive(Serialize, Deserialize, Debug)]
struct Metadata {
    firmware_version: String,
    hardware_id: String,
    calibration_date: String,
    last_maintenance: String,
}

struct SensorDataSubscriber {
    client: Client,
    encoding: String,
    qos: QoS,
    message_count: u32,
    receive_times: Vec<f64>,
}

impl SensorDataSubscriber {
    fn new(broker: String, port: u16, encoding: String, qos: u8) -> Self {
        let mut mqttoptions = MqttOptions::new("rust-subscriber", broker, port);
        mqttoptions.set_keep_alive(std::time::Duration::from_secs(60));
        
        let (client, _eventloop) = Client::new(mqttoptions, 10);
        
        let qos = match qos {
            0 => QoS::AtMostOnce,
            1 => QoS::AtLeastOnce,
            2 => QoS::ExactlyOnce,
            _ => QoS::AtLeastOnce,
        };
        
        Self {
            client,
            encoding,
            qos,
            message_count: 0,
            receive_times: Vec::new(),
        }
    }
    
    fn decode_message(&self, payload: &[u8]) -> Result<SensorData, Box<dyn std::error::Error>> {
        match self.encoding.as_str() {
            "json" => {
                let data: SensorData = serde_json::from_slice(payload)?;
                Ok(data)
            },
            "msgpack" => {
                let data: SensorData = rmp_serde::from_slice(payload)?;
                Ok(data)
            },
            "cbor" => {
                let data: SensorData = ciborium::de::from_reader(payload)?;
                Ok(data)
            },
            "protobuf" => {
                // For protobuf, we would need to convert from protobuf message
                // For now, fall back to JSON
                let data: SensorData = serde_json::from_slice(payload)?;
                Ok(data)
            },
            _ => Err("Unsupported encoding".into()),
        }
    }
    
    fn connect(&mut self, topic: &str) -> Result<(), Box<dyn std::error::Error>> {
        println!("Connecting to MQTT broker at {}:{}...", 
                 self.client.options().broker_address(), 
                 self.client.options().port());
        
        self.client.subscribe(topic, self.qos)?;
        println!("✓ Subscribed to topic: {} (QoS: {:?})", topic, self.qos);
        println!("\nWaiting for messages (Ctrl+C to exit)...\n");
        
        Ok(())
    }
    
    fn handle_message(&mut self, topic: &str, payload: &[u8]) {
        let receive_time = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs_f64();
        
        match self.decode_message(payload) {
            Ok(data) => {
                self.message_count += 1;
                
                // Calculate receive latency if timestamp is available
                let latency = receive_time - data.timestamp;
                self.receive_times.push(latency);
                
                println!("\n[Message {}] Topic: {}", self.message_count, topic);
                println!("  Sensor ID: {}", data.sensor_id);
                println!("  Temperature: {}°C", data.temperature);
                println!("  Humidity: {}%", data.humidity);
                println!("  Pressure: {} hPa", data.pressure);
                println!("  Timestamp: {}", data.timestamp);
                println!("  Receive latency: {:.2}ms", latency * 1000.0);
            },
            Err(e) => {
                eprintln!("Error decoding message: {}", e);
            }
        }
    }
    
    fn disconnect(&self) {
        println!("\n✓ Received {} messages", self.message_count);
        if !self.receive_times.is_empty() {
            let avg_latency = self.receive_times.iter().sum::<f64>() / self.receive_times.len() as f64;
            println!("✓ Average receive latency: {:.2}ms", avg_latency * 1000.0);
        }
        println!("✓ Disconnected");
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args = Args::parse();
    
    println!("=== MQTT Subscriber (Rust) ===");
    println!("Encoding: {}", args.encoding);
    println!("Topic: {}", args.topic);
    println!("QoS: {}", args.qos);
    println!();
    
    let mut subscriber = SensorDataSubscriber::new(
        args.broker,
        args.port,
        args.encoding,
        args.qos,
    );
    
    subscriber.connect(&args.topic)?;
    
    // Create a new client for the event loop
    let mut mqttoptions = MqttOptions::new("rust-subscriber-loop", args.broker, args.port);
    mqttoptions.set_keep_alive(std::time::Duration::from_secs(60));
    
    let (mut client, mut eventloop) = Client::new(mqttoptions, 10);
    client.subscribe(&args.topic, subscriber.qos)?;
    
    loop {
        match eventloop.poll().await {
            Ok(Event::Incoming(Incoming::Publish(publish))) => {
                subscriber.handle_message(&publish.topic, &publish.payload);
            },
            Ok(_) => {},
            Err(e) => {
                eprintln!("Error in event loop: {}", e);
                break;
            }
        }
    }
    
    subscriber.disconnect();
    Ok(())
}
